schnorr_fun/src/frost.rs

```diff
@@ -473,30 +438,50 @@ impl<H: Digest<OutputSize = U32> + Clone, NG: NonceGen> Frost<H, NG> {
 
     /// Generate nonces for creating signatures shares.
     ///
+    /// ⚠ You must use a CAREFULLY CHOSEN nonce rng, see [`FrostKey::gen_nonce_rng`]
+    pub fn gen_nonce<R: RngCore>(&self, nonce_rng: &mut R) -> NonceKeyPair {
+        NonceKeyPair::random(nonce_rng)
+    }
+
+    /// Generate a reusable nonce rng.
+    ///
     /// This method should be used carefully.
-    /// This calls [`NonceKeyPair::generate`] internally with the frost instance's [`NonceGen`].
-    /// See documentation for that for more usage info.
     ///
     /// When choosing a `secret` to use, if you are generating nonces prior to [`KeyGen`] completion,
     /// use the static first coefficient of your polynomial.
     /// Otherwise you can use your secret share of the frost key.
     ///
     /// The application should decide upon a unique `session_id`. If the `NonceGen` of this FROST
     /// instance is `Deterministic` then the `session_id` **must** be unique per signing session.
-    pub fn gen_nonce(
+    pub fn gen_nonce_rng<R: SeedableRng>(
         &self,
         secret: &Scalar,
         session_id: &[u8],
         public_key: Option<Point<impl Normalized>>,
```

Comment:
the main point of this PR from my perspective was so you can pass in `FrostKey<T>` here and get the verification shares out and hash them too. Actually I think strictly speaking if you hash the threshold and all the verification shares that strictly determines the public key anyway (via lagrange interpolatoin).

I think last time we spoke I mentioned that we should just use `ChaCha20Rng`, add a feature flag here and make it concrete. After looking at it I changed my mind. I like what you've done here better. I think you can simplify the type constraint here though. Should just be `R: SeedableRng<Seed=[u8;32]>`